File: R/gBox.R
require(stats)

gBox <- function (model, lags = 1:20, x = eval(model$call$x), method = c("squared", 
    "absolute")[1], plot = TRUE) {
    zlag <- function (x, d = 1) {
        if (d != as.integer(d) || d < 0) 
            stop("d must be a non-negative integer")
        if (d == 0) 
            return(x)
        else return(c(rep(NA, d), rev(rev(x)[-(1:d)])))
    }
    
    kurtosis <- function (x, na.rm = FALSE) {
        var = function(x, ...) {
            mean((x - mean(x, ...))^2)
        }
        if (na.rm) 
            x <- x[!is.na(x)]
        sum((x - mean(x))^4)/(length(x) * var(x, na.rm = na.rm)^2) - 
            3
    }
    
    acf <- function (x, lag.max = NULL, type = c("correlation", "covariance", 
        "partial")[1], plot = TRUE, na.action = na.fail, demean = TRUE, 
        drop.lag.0 = TRUE, ...) {
        acf.out <- stats::acf(x = x, lag.max = lag.max, type = type, 
            plot = F, na.action = na.action, demean = demean, ...)
        acf.out$series <- deparse(substitute(x))
        if (drop.lag.0) {
            if (type == "correlation") {
                acf.out$acf = acf.out$acf[-1, , , drop = FALSE]
                acf.out$lag = acf.out$lag[-1, , , drop = FALSE]
            }
        }
        if (plot) {
            plot1.acf(acf.out, ...)
            return(invisible(acf.out))
        }
        else return(acf.out)
    }
    
    gBox.test <- function(model, lag, x) {
        p = model$order[1]
        q = model$order[2]
        start = 1 + q
        beta = model$coef[(start + 1):(start + p)]
        r = x
        epsilon = residuals(model)
        M = NULL
        i = 1
        while (i <= q) {
            M = cbind(M, zlag(r^2, i))
            i = i + 1
        }
        i = 1
        h = model$fit[, 1]
        h = h * h
        while (i <= p) {
            M = cbind(M, zlag(h, i))
            i = i + 1
        }
        sigma2 = mean(r^2)
        M[is.na(M)] = sigma2
        if (p > 0) 
            M = apply(M, 2, filter, filter = beta, method = "recursive", 
                sides = 1, init = rep(sigma2, length(beta)))
        if (p > 0) 
            M = cbind(1/(1 - sum(beta)), M)
        else M = cbind(1, M)
        M = apply(M, 2, function(x) {
            x/h
        })
        M[is.na(M)] = 0
        E = NULL
        for (i in 1:lag) {
            E = cbind(E, zlag(epsilon^2, i) - mean(epsilon^2, 
                na.rm = TRUE))
        }
        E[is.na(E)] = 0
        J = t(E) %*% M/length(h)
        k = kurtosis(epsilon, na.rm = TRUE) + 2
        Lambda = 2 * solve(t(M) %*% M)
        H = -J %*% Lambda %*% t(J)/(2 * k) * length(h)
        diag(H) = 1 + diag(H)
        acf1 = matrix(as.vector(acf(epsilon^2, lag.max = lag, 
            na.action = na.omit, plot = FALSE)$acf), ncol = 1)
        stat = as.numeric(t(acf1) %*% solve(H) %*% acf1 * length(h))
        p.value = 1 - pchisq(stat, lag)
        return(invisible(list(p.value = p.value, H = H, Lambda = Lambda, 
            J = J, stat = stat, acf = acf1, n = length(h))))
    }
    
    gBox1.test <- function(model, lag, x) {
        p = model$order[1]
        q = model$order[2]
        start = 1 + q
        beta = model$coef[(start + 1):(start + p)]
        r = x
        epsilon = residuals(model)
        M = NULL
        i = 1
        while (i <= q) {
            M = cbind(M, zlag(r^2, i))
            i = i + 1
        }
        i = 1
        h = model$fit[, 1]
        h = h * h
        while (i <= p) {
            M = cbind(M, zlag(h, i))
            i = i + 1
        }
        sigma2 = mean(r^2)
        M[is.na(M)] = sigma2
        if (p > 0) 
            M = apply(M, 2, filter, filter = beta, method = "recursive", 
                sides = 1, init = rep(sigma2, length(beta)))
        if (p > 0) 
            M = cbind(1/(1 - sum(beta)), M)
        else M = cbind(1, M)
        M = apply(M, 2, function(x) {
            x/h
        })
        M[is.na(M)] = 0
        E = NULL
        tau = mean(abs(epsilon), na.rm = TRUE)
        nu = mean(abs(epsilon)^3, na.rm = TRUE)
        for (i in 1:lag) {
            E = cbind(E, zlag(abs(epsilon), i) - tau)
        }
        E[is.na(E)] = 0
        J = t(E) %*% M/length(h)
        k = kurtosis(epsilon, na.rm = TRUE) + 2
        Lambda = 2 * solve(t(M) %*% M) * k/2
        H = J %*% Lambda %*% t(J) * length(h) * (k/2 * tau^2/4 - 
            tau * (nu - tau)/2)/(1 - tau^2)^2
        diag(H) = 1 + diag(H)
        acf1 = matrix(as.vector(acf(abs(epsilon), lag.max = lag, 
            na.action = na.omit, plot = FALSE)$acf), ncol = 1)
        stat = as.numeric(t(acf1) %*% solve(H) %*% acf1 * length(h))
        p.value = 1 - pchisq(stat, lag)
        return(invisible(list(p.value = p.value, H = H, Lambda = Lambda, 
            J = J, stat = stat, acf = acf1, n = length(h))))
    }
    
    pv = NULL
    for (lag in lags) {
        new = switch(method, squared = gBox.test(model, lag, 
            x)$p.value, absolute = gBox1.test(model, lag, x)$p.value)
        pv = c(pv, new)
    }
    if (plot) {
        plot(x = 1:20, y = pv, ylim = c(0, 1), xlab = "lag", 
            ylab = "p-value", main = "")
        abline(h = 0.05, lty = 2)
    }
    invisible(list(pvalue = pv, lags = lags, method = method, 
        x = x))
}

File: Python/gBox.py
import numpy as np
import pandas as pd

from statsmodels.tsa.stattools import acf
from scipy.signal import lfilter, lfiltic
from scipy.stats import chi2

import matplotlib.pyplot as plt


def gBox(fitted_model, lags=None, x=None, method="squared", plot=True, **kwargs):
    """ Adapted from R's TSA library function gBox, but to be used with
    models returned from Python's arch library. """
    
    def zlag(x, d=1):
        return np.array(pd.Series(x).shift(d))

    def kurtosis(x, na_rm=False):
        if na_rm:
            x = x[~np.isnan(x)]
        return np.sum((x - np.mean(x))**4) / (len(x) * np.var(x)**2) - 3

    def gBox_test(fitted_model, lag, x):
        p = fitted_model.model.volatility.p
        q = fitted_model.model.volatility.q
        beta = np.array([fitted_model.params['beta[' + str(i + 1) +']'] for i in range(q)])
        #x = fitted_model.model.y
        epsilon = zlag(fitted_model.resid, d=max(p, q))
        M = np.empty((len(x), p+q))
        for i in range(q):
            M[:, i] = zlag(x**2, i+1)
        h = zlag(fitted_model.conditional_volatility, max(p, q))**2
        for i in range(p):
            M[:, q + i] = zlag(h, i+1)
        sigma2 = np.mean(x**2)
        np.nan_to_num(M, nan=sigma2, copy=False)

        if p > 0:
            a, b = np.r_[1, -beta], np.r_[1.0]
            zi = lfiltic(b=b, a=a, y=np.ones(len(beta)) * sigma2)
            def beta_filter(x):
                y, zf = lfilter(b=b, a=a, x=x, zi=zi)
                return y

            M_swap = np.apply_along_axis(func1d=beta_filter, axis=0, arr=M)
            M = np.empty((len(x), p+q+1))
            M[:, 0] = 1 / (1 - np.sum(beta))
            M[:, 1:] = M_swap
        else:
            M_swap = M.copy()
            M = np.empty((len(x), p+q+1))
            M[:, 0] = 1
            M[:, 1:] = M_swap

        M = np.apply_along_axis(func1d=lambda c : c / h, axis=0, arr=M)
        np.nan_to_num(M, nan=0, copy=False)

        E = np.empty((lag, len(x)))
        e2 = epsilon**2
        e2_mean = np.nanmean(e2)
        for i in range(lag):
            E[i, :] = zlag(e2, i+1) - e2_mean
        np.nan_to_num(E, nan=0, copy=False)

        J = (E @ M)/len(h)
        k = kurtosis(epsilon, na_rm=True)
        Lambda = 2 * np.linalg.inv(M.T @ M)
        H = -J @ Lambda @ J.T/(2 * k) * len(h)
        for i in range(lag):
            H[i, i] += 1
        acf1 = np.resize(acf(e2, nlags=lag, fft=False, missing='drop')[1:], (lag, 1))
        stat = (acf1.T @ np.linalg.inv(H) @ acf1 * len(h))[0, 0]
        p_value = 1 - chi2.cdf(x=stat, df=lag)

        return {
            'p_value': p_value,
            'H': H,
            'Lambda': Lambda,
            'J': J,
            'stat': stat,
            'acf': acf1,
            'n': len(h)
        }

    def gBox1_test(fitted_model, lag, x):
        p = fitted_model.model.volatility.p
        q = fitted_model.model.volatility.q
        beta = np.array([fitted_model.params['beta[' + str(i + 1) +']'] for i in range(q)])
        #x = model.model.y
        epsilon = zlag(fitted_model.resid, d=max(p, q))
        M = np.empty((len(x), p+q))
        for i in range(q):
            M[:, i] = zlag(x**2, i+1)
        h = zlag(fitted_model.conditional_volatility, max(p, q))**2
        for i in range(p):
            M[:, q + i] = zlag(h, i+1)
        sigma2 = np.mean(x**2)
        np.nan_to_num(M, nan=sigma2, copy=False)

        if p > 0:
            a, b = np.r_[1, -beta], np.r_[1.0]
            zi = lfiltic(b=b, a=a, y=np.ones(len(beta)) * sigma2)
            def beta_filter(x):
                y, zf = lfilter(b=b, a=a, x=x, zi=zi)
                return y

            M_swap = np.apply_along_axis(func1d=beta_filter, axis=0, arr=M)
            M = np.empty((len(x), p+q+1))
            M[:, 0] = 1 / (1 - np.sum(beta))
            M[:, 1:] = M_swap
        else:
            M_swap = M.copy()
            M = np.empty((len(x), p+q+1))
            M[:, 0] = 1
            M[:, 1:] = M_swap

        M = np.apply_along_axis(func1d=lambda c : c / h, axis=0, arr=M)
        np.nan_to_num(M, nan=0, copy=False)

        e_abs = np.abs(epsilon)
        tau = np.nanmean(e_abs)
        nu = np.nanmean(e_abs**3)

        E = np.empty((lag, len(x)))

        for i in range(lag):
            E[i, :] = zlag(e_abs, i+1) - tau
        np.nan_to_num(E, nan=0, copy=False)

        J = (E @ M)/len(h)
        k = kurtosis(epsilon, na_rm=True)
        Lambda = 2 * np.linalg.inv(M.T @ M)

        H = (J @ Lambda @ J.T) * len(h) * (k/2 * tau**2/4 - tau * (nu - tau)/2)/(1 - tau**2)**2

        for i in range(lag):
            H[i, i] += 1
        acf1 = np.resize(acf(e_abs, nlags=lag, fft=False, missing='drop')[1:], (lag, 1))
        stat = (acf1.T @ np.linalg.inv(H) @ acf1 * len(h))[0, 0]
        p_value = 1 - chi2.cdf(x=stat, df=lag)

        return {
            'p_value': p_value,
            'H': H,
            'Lambda': Lambda,
            'J': J,
            'stat': stat,
            'acf': acf1,
            'n': len(h)
        }
    
    
    if x is None:
        x = fitted_model.model.y
    if lags is None:
        lags = np.arange(1, 21)
    
    get_p_value = None
    if method == "squared":
        get_p_value = lambda lag : gBox_test(fitted_model, lag, x)['p_value']
    elif method == "absolute":
        get_p_value = lambda lag : gBox1_test(fitted_model, lag, x)['p_value']
    else:
        raise "Unknown method: " + str(method)
    
    results = np.empty(len(lags))
    for i, lag in enumerate(lags):
        results[i] = get_p_value(lag)
    
    if plot:
        ax = plt.gca()
        kwargs = {'marker': 'o', 'markersize': 5, 'linestyle': None}
        ax.margins(.05)
        ax.plot(lags, results, **kwargs)
        ax.set_ylim(-0.05, 1.05)
        ax.axhline(0.05, xmin=np.min(lags)-1, xmax=np.max(lags)+1, color='red', linestyle=':')
        ax.set_xlabel('Lag')
        ax.set_ylabel('P-value')
    
    return {
        'lags': lags,
        'results': results
    }